# Background information responding to Gihawi et al. 2023 bioRxiv 

This Github repository is a succint response to the _second_ [preprint](https://www.biorxiv.org/content/10.1101/2023.07.28.550993v1) posted by Gihawi et al. on July 31, 2023 regarding our [2020 _Nature_ paper](https://www.nature.com/articles/s41586-020-2095-1), and it only addresses whether their data tables show cancer type-specific signals. 

Approximately 5 months ago, we previously posted our [extensive rebuttal](https://www.biorxiv.org/content/10.1101/2023.02.10.528049v1) of the Gihawi and colleagues' [_first_ preprint](https://www.biorxiv.org/content/10.1101/2023.01.16.523562v1), which contains many of the same topics as their second preprint. We further note that many of their concerns were concurrently addressed in our peer-reviewed [2022 _Cell_ paper](https://www.cell.com/cell/pdf/S0092-8674(22)01127-8.pdf), which employed a wide variety of updated methods (e.g., newer microbial reference database, alignment methods, host depletion, machine learning, de novo metagenome assembly, etc.) compared to the 2020 paper while making the same conclusion: microbiomes in tumors and blood are cancer type-specific.

# (1) Re-analysis of raw data provided by Gihawi et al. 2023 bioRxiv still show cancer type-specific microbiomes

We asked whether their raw data on 3 TCGA cancer types (bladder [BLCA], breast [BRCA], head and neck [HNSC]) support cancer type-specific microbiomes in tumors and blood, which was the central conclusion of our 2020 paper. For reference, their data were downloaded from the Github link ([https://github.com/yge15/Cancer_Microbiome_Reanalyzed](https://github.com/yge15/Cancer_Microbiome_Reanalyzed)) listed in their Methods section.

Briefly, the following was done:
- Their data were loaded alongside metadata from the original [Poore et al. 2020 Nature](https://www.nature.com/articles/s41586-020-2095-1) paper
- The 3 cancer type tables containing raw counts were merged. Conservatively, we retained only overlapping genera between the 3 tables. However, we removed all counts associated with the _Homo_ genus (i.e., human data) to focus just on the microbial differences. As noted by Gihawi et al., their database for KrakenUniq only contained complete genomes of microbes.
- We separated the count and metadata into subsets with a single sequencing center (e.g., Harvard Medical School [HMS]), sequencing platform (Illumina HiSeq), and experimental strategy (WGS) in order to avoid needing to apply batch correction. All subsequent analyses were based on these raw data subsets without batch correction.
- We applied beta and alpha diversity analyses to these subsets using [Qiime 2](https://qiime2.org/) (see code within the `Qiime` folder). To avoid rarefaction for the beta diversity analyses, we employed a robust Aitchison approach called RPCA ([Martino et al. 2019](https://journals.asm.org/doi/10.1128/msystems.00016-19)). We also applied more traditional beta-diversity measures (e.g., Bray-Curtis, Jaccard) and calculated alpha diversity using data rarefied to the 1st quartile of sample read counts for each subset.
- After finding that the cancer type separation was significant, we performed multi-class and two-class machine learning within HMS primary tumor (PT) samples and HMS blood derived normal (BDN) samples.

The figure summarizing the findings and its associated caption is below, demonstrating the raw data still provide cancer type specificity (**Figure 1**). All analyses necessary to generate these plots have code contained within this repository or can be done without code using [https://view.qiime2.org/](https://view.qiime2.org/).

![Alt text](Figures/assembled-figure.jpg?raw=true "Title")

**Figure 1: Raw data provided by Gihawi et al. 2023 bioRxiv reveal cancer type-specific microbiomes.**  
**(A)** The breakdown of samples derived from TCGAâ€™s Harvard Medical School (HMS), which contributed the most samples of any sequencing center among their samples (284 of 728, 39.0%). To avoid the need for batch correction, samples were analyzed using raw data from Gihawi et al. 2023 in subsets from a single sequencing center (here, HMS), sequencing platform (Illumina HiSeq), and experimental strategy (WGS). The data were not transformed in any way other than removing the Homo genus counts since it was included in their KrakenUniq mappings.  
**(B)** Aitchison-based Principal Coordinates Analysis (RPCA) on the raw data, excluding Homo counts, across all primary tumors (PTs) from HMS. RPCA does not require rarefaction and none was used. PERMANOVA with 999 iterations was run to estimate effect sizes of cancer type separation, which was significant (inset values, p=0.001). Colors reflect cancer types and are uniform across all subpanels of this figure.  
**(C)** After RPCA, Shannon entropies were calculated on all PT samples from HMS using raw data (excluding Homo counts) rarefied to the first quartile of sample counts (600 reads/sample). A Kruskal-Wallis test showed significant differences in Shannon diversity among cancer types (p=1.4e-9).  
**(D)** RPCA was then applied to all blood derived normal (BDN) samples from HMS using raw data (excluding Homo counts). RPCA does not require rarefaction and none was used. PERMANOVA with 999 iterations was run to estimate effect sizes of cancer type separation, which was significant (inset values, p=0.001).  
**(E)** After RPCA, Shannon entropies were calculated on all BDN samples from HMS using raw data (excluding Homo counts) rarefied to the first quartile of sample counts (400 reads/sample). A Kruskal-Wallis test showed significant differences in Shannon diversity among cancer types (p=3.7e-7).  
**(F)** Having found cancer type-specific differences, we evaluated if multiclass machine learning could discriminate between cancer types using the raw data (excluding Homo counts) from all HMS PT samples. Gradient boosting machines were applied with 10-fold cross-validation such that every sample was left out once, and their predictions were used to generate a confusion matrix. The mean balanced accuracy was 97.59% in comparison to the no information rate (NIR) of 54.84% (p<2.2e-16). 
**(G)** 10-fold cross-validation using gradient boosting machines on HMS BDN samples. The balanced accuracy was 92.8% in comparison to the NIR of 80% (p=4.4e-5).  

# (2) Subsetting the raw data provided by Gihawi et al. 2023 bioRxiv to known tumor-resident bacteria and fungi from the Weizmann Institute still show cancer type-specific microbiomes

Above, we considered the raw data as-is from Gihawi and colleagues. We next evaluated whether the predictive modeling results would be replicated if subsetting the microbial genera to only those identified in a highly-decontaminated set of independent tumors processed by the Weizmann Institute of Science (WIS) (see [Nejman et al. 2020 Science](https://www.science.org/doi/10.1126/science.aay9189), [Narunsky-Haziza et al. 2022 Cell](https://www.cell.com/cell/pdf/S0092-8674(22)01127-8.pdf)). We note these two studies together ran >1100 experimental contamination controls in parallel to the tumor samples, thereby providing strong support for the taxa that persisted. For further background, we also refer to the [answer in this Github comment](https://github.com/gregpoore/tcga_rebuttal/issues/1#issuecomment-1663410443). After importing the 'hit list' of bacterial and fungal genera and intersecting these genera with the raw data from Gihawi and colleagues (n=149 remaining genera, 17.6% of total), we then repeated the same machine learning analyses as done above, indeed finding similarly strong cancer type discrimination (**Figure 2**). Thus, using highly-decontaminated microbial genera from an independent set of tumors, we find that the raw data from Gihawi and colleagues still shows cancer type-specificity in TCGA.

The figure summarizing the findings and its associated caption is below. The code necessary to generate these plots is contained within `tcga_gihawhi_rebuttal_WIS_subset_3Aug23.R`.

![image](https://user-images.githubusercontent.com/7695851/258028437-e6fbd5c4-4d03-4311-97f9-ed78629701fd.png)  
**Figure 2: Known tumor-resident genera from the Weizmann Institute of Science (WIS) in data from Gihawi et al. 2023 bioRxiv reveal cancer type-specific microbiomes in TCGA tumors and blood.**  
**(A)** After subsetting to WIS-overlapping genera (n=149 genera), we evaluated if multiclass machine learning could discriminate between cancer types using the raw data from all HMS PT samples. Gradient boosting machines were applied with 10-fold cross-validation such that every sample was left out once, and their predictions were used to generate a confusion matrix. The mean balanced accuracy was 93.62% in comparison to the no information rate (NIR) of 54.84% (p<2.2e-16).  
**(B)** After subsetting to WIS-overlapping genera (n=149 genera), 10-fold cross-validation using gradient boosting machines was applied on HMS BDN samples. The balanced accuracy was 88.82% in comparison to the NIR of 80% (p=4.4e-5).


